Add --version

IsmayilMirzali opened the pull request:
This adds the option to use ```nmstatectl version```.
@ffmancera

ffmancera commented on nmstatectl/nmstatectl.py:
```diff
@@ -157,6 +158,17 @@ def setup_subcommand_show(subparsers):
 
+def setup_subcommand_displayversion(subparsers):
+    parser_display_version = subparsers.add_parser(
+        "version",
```
It is not possible to rename this to "--version"? Just asking I didn't try it.

IsmayilMirzali replied:
I get a message saying it expects an argument, I can definitely change it but I don't quite know how to play around that arg requirement.

IsmayilMirzali replied:
I get an error like this:
``` bash
[root@16e920d9c918 nmstate-workspace]# nmstatectl --version
usage: nmstatectl [-h] {commit,edit,rollback,set,show,--version} ...
nmstatectl: error: unrecognized arguments: --version
```

ffmancera replied:
I see, let's keep it as `nmstatectl version`. @tyll What do you think?

tyll replied:
For `--version` it needs something like this:

```
parser.add_argument("--version")
...
if args.version:
   print version
else:
    return args.func(args)
```

`nmstatectl version` makes probably more sense but I would assume everyone would expect `nmstatectl --version` to work, too.

ffmancera replied:
Yes, it would be nice to have both. Thanks!

IsmayilMirzali replied:
I'll try my best to figure out how to do it with ```nmstatectl --version```, since that's also the way I'd expect things to work. I was reading argparse documentation but the way the main function is structured makes it harder for me follow. Are we sure this is something that can be implemented within ```nmstatectl.py``` ?

tyll replied:
Did you try the approach I outlined in https://github.com/nmstate/nmstate/pull/920#discussion_r390406218?

IsmayilMirzali replied:
@tyll Yep, finally made it work, thanks. Just had to check documentation to make it not require extra arguments past ```--version```. Will update the PR with the changes and the integration test I wrote.

ffmancera commented on nmstatectl/nmstatectl.py:
```diff
@@ -51,6 +51,7 @@ def main():
     setup_subcommand_rollback(subparsers)
     setup_subcommand_set(subparsers)
     setup_subcommand_show(subparsers)
+    setup_subcommand_displayversion(subparsers)
```
Please, rename this to `setup_subcommand_version`.

IsmayilMirzali commented:
@ffmancera Added the integration test and implemented ```--version``` instead. Works as intended on my local copy and passed the test as well.

ffmancera commented:
> @ffmancera Added the integration test and implemented `--version` instead. Works as intended on my local copy and passed the test as well.

That is very good. I do not see the commit containing the integration tests. In addition it would be nice to have both, `--version` and `nmstatectl version`. Could you do it?

Please, when working on the suggestions I recommend you to amend the changes to the commit so this way you are going to have a single commit. I recommend you to take a look to [Pro Git book](https://git-scm.com/book/en/v2). 

Thank you!

IsmayilMirzali commented:
> > @ffmancera Added the integration test and implemented `--version` instead. Works as intended on my local copy and passed the test as well.
> 
> That is very good. It would be nice to have both, `--version` and `nmstatectl version`. Could you do it?
> 
> Please, when working on the suggestions I recommend you to amend the changes to the commit so this way you are going to have a single commit. I recommend you to take a look to [Pro Git book](https://git-scm.com/book/en/v2).
> 
> Thank you!

Added ```nmstatectl version``` back in. Tried amending the commits to make it one commit but didn't quite work out.

ffmancera commented:
> > > @ffmancera Added the integration test and implemented `--version` instead. Works as intended on my local copy and passed the test as well.
> > 
> > 
> > That is very good. It would be nice to have both, `--version` and `nmstatectl version`. Could you do it?
> > Please, when working on the suggestions I recommend you to amend the changes to the commit so this way you are going to have a single commit. I recommend you to take a look to [Pro Git book](https://git-scm.com/book/en/v2).
> > Thank you!
> 
> Added `nmstatectl version` back in. Tried amending the commits to make it one commit but didn't quite work out.

Good! Please, try to squash the commits in order to have a single commit. In addition it would be nice to have also an integration test for `nmstatectl_version`. I suggest you the following names `test_version_command` and `test_version_argument`.

Please make sure all the tests are passing, it seems that linter and formatter are complaining about something. Looks quite good, thank you!

IsmayilMirzali commented:
Closing as this PR is very messy. Will reopen a clean one.

tyll commented:
> Closing as this PR is very messy. Will reopen a clean one.

What is the problem with this PR that you would like to solve by opening a new one?

IsmayilMirzali commented:
> > Closing as this PR is very messy. Will reopen a clean one.
> 
> What is the problem with this PR that you would like to solve by opening a new one?

The commits felt kinda messy and I had trouble with squashing them properly into 1 commit. I deleted my old fork and implemented the changes on a new fork and opened PR from that new fork.

tyll commented:
> > > Closing as this PR is very messy. Will reopen a clean one.
> > 
> > 
> > What is the problem with this PR that you would like to solve by opening a new one?
> 
> The commits felt kinda messy and I had trouble with squashing them properly into 1 commit. I deleted my old fork and implemented the changes on a new fork and opened PR from that new fork.

The discussion is important to know why we got to this decision. It is possible to squash the commits and update this branch. Please do that in the future. Seefor example https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History for reference.
